src/Physics/Multinucleon/XSection/MECScaleVsW.cxx

```diff
@@ -0,0 +1,170 @@
+//_________________________________________________________________________
+/*
+ Copyright (c) 2003-2020, The GENIE Collaboration
+ For the full text of the license visit http://copyright.genie-mc.org
+
+ Original code contributed by J.Tena and M.Roda
+ Substantial code refactorizations by the core GENIE group.
+*/
+//_________________________________________________________________________
+
+#include "Physics/Multinucleon/XSection/MECScaleVsW.h"
+#include "Framework/Utils/StringUtils.h" 
+#include "Framework/ParticleData/PDGCodes.h"
+#include "Framework/ParticleData/PDGLibrary.h"
+
+using namespace genie;
+
+//_________________________________________________________________________
+MECScaleVsW::MECScaleVsW() : 
+  XSecScaleI("genie::MECScaleVsW")
+{
+  
+}
+//_________________________________________________________________________
+MECScaleVsW::MECScaleVsW(string config) : 
+  XSecScaleI("genie::MECScaleVsW",config) 
+{
+  
+}
+//_________________________________________________________________________
+MECScaleVsW::~MECScaleVsW()
+{
+
+}
+//_________________________________________________________________________
+double MECScaleVsW::GetScaling( const Interaction & interaction ) const
+{
+  double Q0 = interaction.Kine().GetKV(kKVQ0) ; 
+  double Q3 = interaction.Kine().GetKV(kKVQ3) ; 
+
+  return GetScaling( Q0, Q3 ) ; 
+}
+//_________________________________________________________________________
+double MECScaleVsW::GetScaling( const double Q0, const double Q3 ) const
+{
+  // Get the vectors that include the kinematic limits of W for a given event:
+  MECScaleVsW::weight_type_map weight_map = GetMapWithLimits( Q0, Q3 ) ;
+  
+  // The Scaling is done using the "experimenter's W", which assumes a single nucleon
+  // See motivation in : https://arxiv.org/pdf/1601.02038.pdf
+  // Calculate event W:
+  static double Mn = ( PDGLibrary::Instance()->Find(kPdgProton)->Mass() + PDGLibrary::Instance()->Find(kPdgNeutron)->Mass() ) * 0.5 ;  // Nucleon mass
+  double W = sqrt( pow(Mn,2) + 2*Mn*Q0 - pow(Q3,2) + pow(Q0,2) ) ;
+
+  // Calculate scaling:
```

Comment:
Some optimisations and ideas.
1. Use two iterators instead of one. `it_max` and `it_min`
2. Before the loop check if W is outside the range and in that case return the default weight
3. the loop is done in a way that you divide the distance between the iterators in 2 every step of the loop.